metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -273,390 +273,397 @@ class MetalsLanguageServer(
-    )
-    shellRunner = register(
-      new ShellRunner(languageClient, () => userConfig, time, statusBar)
-    )
-    bloopInstall = new BloopInstall(
-      workspace,
-      languageClient,
-      buildTools,
-      tables,
-      shellRunner
-    )
-    bspConfigGenerator = new BspConfigGenerator(
-      workspace,
-      languageClient,
-      buildTools,
-      shellRunner
-    )
-    newProjectProvider = new NewProjectProvider(
-      languageClient,
-      statusBar,
-      clientConfig,
-      shellRunner,
-      clientConfig.icons,
-      workspace
-    )
-    bloopServers = new BloopServers(
-      workspace,
-      buildClient,
-      languageClient,
-      tables,
-      clientConfig.initialConfig
-    )
-    bspServers = new BspServers(
-      workspace,
-      charset,
-      languageClient,
-      buildClient,
-      tables,
-      bspGlobalDirectories,
-      clientConfig.initialConfig
-    )
-    buildToolSelector = new BuildToolSelector(
-      languageClient,
-      tables
-    )
-    bspConnector = new BspConnector(
-      bloopServers,
-      bspServers,
-      buildTools,
-      languageClient,
-      tables,
-      () => userConfig,
-      statusBar
-    )
-    semanticdbs = AggregateSemanticdbs(
-      List(
-        fileSystemSemanticdbs,
-        interactiveSemanticdbs
-      )
-    )
-    definitionProvider = new DefinitionProvider(
-      workspace,
-      mtags,
-      buffers,
-      definitionIndex,
-      semanticdbs,
-      warnings,
-      () => compilers,
-      remote
-    )
-    formattingProvider = new FormattingProvider(
-      workspace,
-      buffers,
-      () => userConfig,
-      languageClient,
-      clientConfig,
-      statusBar,
-      clientConfig.icons,
-      Option(params.getWorkspaceFolders) match {
-        case Some(folders) =>
-          folders.asScala.map(_.getUri.toAbsolutePath).toList
-        case _ =>
-          Nil
-      },
-      tables
-    )
-    newFileProvider = new NewFileProvider(
-      workspace,
-      languageClient,
-      packageProvider,
-      () => focusedDocument
-    )
-    referencesProvider = new ReferenceProvider(
-      workspace,
-      semanticdbs,
-      buffers,
-      definitionProvider,
-      remote
-    )
-    implementationProvider = new ImplementationProvider(
-      semanticdbs,
-      workspace,
-      definitionIndex,
-      buildTargets,
-      buffers,
-      definitionProvider
-    )
+    // NOTE: we purposefully don't check workspaceFolders here
+    // since Metals technically doesn't support it. Once we implement
+    // https://github.com/scalameta/metals-feature-requests/issues/87 we'll
+    // have to change this.
+    val root =
+      Option(params.getRootUri()).orElse(Option(params.getRootPath()))
 
-    supermethods = new Supermethods(
-      languageClient,
-      definitionProvider,
-      implementationProvider
-    )
+    root match {
+      case None =>
+        languageClient.showMessage(Messages.noRoot)
```

Comment:
This pr looks huge because of the shift, but really this is the only difference. If there is no root, we just show the message, and then don't replace the providers.

Reply:
![image](https://user-images.githubusercontent.com/691940/104718269-e9577d00-572a-11eb-95a6-f7de81aece6e.png)

This helps a lot with the review

Reply:
ahhhh TIL! I've never used this before!